amethyst_network/src/systems/network_client.rs

```diff
@@ -0,0 +1,135 @@
+    //omg unreadable plz enjoy code owo
+    fn run(&mut self, (mut events,): Self::SystemData) {
+        let mut buf = [0; 2048];
+        loop {
+            match self.socket.recv_from(&mut buf) {
+                //Data received
+                Ok((amt, src)) => {
+                    //Are we connected to anything?
+                    if self.connection.is_some() {
+                        //Was it sent by connected server, and are we still connected to it?
+                        if src == self.connection.as_ref().unwrap().target
+                            && (self.connection.as_ref().unwrap().state
+                                == ConnectionState::Connected
+                                || self.connection.as_ref().unwrap().state
+                                    == ConnectionState::Connecting)
+                        {
+                            let net_event = self.deserialize_event(&buf[..amt]);
+                            match net_event {
+                                Ok(ev) => {
+                                    let owned_event = NetOwnedEvent {
+                                        event: ev.clone(),
+                                        owner: self.connection.as_ref().unwrap().clone(),
+                                    };
+                                    events.single_write(owned_event);
+                                    match T::custom_to_base(ev) {
+                                        Some(NetEvent::Connected) => {
+                                            self.connection.as_mut().unwrap().state =
+                                                ConnectionState::Connected;
+                                            println!(
+                                                "Remote ({:?}) accepted connection request.",
+                                                src
+                                            );
+                                        }
+                                        //Could be handled differently by the user, say by reconnecting to a fallback server.
+                                        Some(NetEvent::ConnectionRefused { reason }) => {
+                                            self.connection = None;
+                                            println!("Connection refused by server: {}", reason);
+                                        }
+                                        Some(NetEvent::Disconnected { reason }) => {
+                                            self.connection = None;
+                                            println!("Disconnected from server: {}", reason);
+                                        }
+                                        _ => {} //Other systems will handle the rest of the stuff
+                                    }
+                                }
+                                Err(e) => println!("Failed to read network event: {}", e),
```

Comment:
It may fail because not whole event was sent.
Protocol should be designed in such a way that receiver may know length of the packet from first few bytes.

Reply:
> Protocol should be designed in such a way that receiver may know length of the packet from first few bytes.

Yes!

Reply:
Also receiver must have a way to find out that received datagram doesn't contain first bytes of the packet. Even in case when those bytes are lost in netherstorm.
There are multiple possible scenarios of failure with different probability.
Protocol must handle 3 sigma of them.

Reply:
`Protocol should be designed in such a way that receiver may know length of the packet from first few bytes.`
https://tr1.cbsistatic.com/hub/i/2015/06/03/59843f0d-0987-11e5-940f-14feb5cc3d2a/r00220010702mul01_03.gif

Reply:
I entirely agree with that, I tended to always do either a 4-byte size prefix to every packet (which was the size of the next N amount of bytes) or I used a packed integer to represent it (0-127 in first byte, if first byte MSB was 1 then it expanded to 2 bytes with 14 bits of information, and repeated out like that).  After the size byte(s) it grabbed the next N amount of bytes and I used either 1/2/4/8 byte or variable packed byte 'type' ID at the start of that chunk.  It makes making bouncers and repeaters far easier and allowing the protocol to pass through for unknown versioned elements when set up like that thus making it more easily forward and backward safe.